### PR TITLE
Automatic emap in HCAL hardcode conditions

### PR DESCRIPTION
**Pull request by kpedro88**
This PR introduces an algorithm to assign ElectronicsIds to any collection of HCAL cells (regardless of topology) and thereby create an ElectronicsMap automatically.

There are also a few changes to the interim customizations for HCAL development, which will be removed when the full workflow is ready.

I also fixed the ptr copy problem in `HcalElectronicsMap` noted by @slava77.

@deguio looked at the DQM output with the automatic emap (following [HcalPhase1SoftwareSimulationRecipe](https://twiki.cern.ch/twiki/bin/viewauth/CMS/HcalPhase1SoftwareSimulationRecipe), using the `-v` option) and saw sensible results.

Still to be done in subsequent PRs:
- Remove use of mutables from `HcalElectronicsMap`
- Adapt offline DQM code to monitor QIE10/11 digis
- Adapt validation code to handle new depths and QIE10/11 digis
- Figure out how to handle ElectronicsId and emap for fine-segmented BH, which has more than 2^19 cells

### CODE REVIEW

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by civanch** [comment]
+1

**Comment by mmusich** [comment]
+1 for changes in `HcalDbHardCode`.
@ggovi the change in CondFormat requires your attention

**Comment by kpedro88** [comment]
rebased...

@cmsbuild please test

**Comment by civanch** [comment]
Please test

**Comment by mmusich** [comment]
+1

**Comment by ggovi** [comment]
+1

**Comment by civanch** [comment]
+1

File: SLHCUpgradeSimulations/Configuration/python/HCalCustoms.py [hunk 1]
```diff
@@ -93,7 +93,13 @@ def customise_Hcal2017(process):
         process=customise_mixing(process)
     if hasattr(process,'simHcalTriggerPrimitiveDigis'):
         process.simHcalTriggerPrimitiveDigis.upgradeHF = cms.bool(True)
-    
+    if hasattr(process,'dqmoffline_step'):
```
Comment (davidlange6) [inline SLHCUpgradeSimulations/Configuration/python/HCalCustoms.py]: probably more reliable just to check hasattr(process,'digiTask' ) but I presume this is all going away in favor of eras soon...

**Comment by davidlange6** [comment]
+1

File: SLHCUpgradeSimulations/Configuration/python/HCalCustoms.py (hunk 1 above)
Comment (kpedro88) [inline SLHCUpgradeSimulations/Configuration/python/HCalCustoms.py]: Yes, HCalCustoms is just being used for development right now. Hopefully everything will be moved to Eras by pre12.

**Pull request merged**